src/dispatch/batch.rs

```diff
@@ -0,0 +1,418 @@
+use crate::{
+    dispatch::Dispatcher, world::ResourceId, Accessor, AccessorCow, DynamicSystemData, RunningTime,
+    System, SystemData, World,
+};
+
+/// The `BatchAccessor` is used to notify the main dispatcher of the read and
+/// write resources of the `System`s contained by the `Batch` ("sub systems").
+#[derive(Debug)]
+pub struct BatchAccessor {
+    reads: Vec<ResourceId>,
+    writes: Vec<ResourceId>,
+}
+
+impl BatchAccessor {
+    /// Create the `BatchAccessor`
+    pub fn new(reads: Vec<ResourceId>, writes: Vec<ResourceId>) -> Self {
+        BatchAccessor { reads, writes }
+    }
+}
+
+impl Accessor for BatchAccessor {
+    fn try_new() -> Option<Self> {
+        None
+    }
+
+    fn reads(&self) -> Vec<ResourceId> {
+        self.reads.clone()
+    }
+
+    fn writes(&self) -> Vec<ResourceId> {
+        self.writes.clone()
+    }
+}
+
+/// The `BatchUncheckedWorld` wrap an instance of the world.
+/// It's safe to be used only in the context of the `Batch`.
+pub struct BatchUncheckedWorld<'a>(pub &'a World);
+
+impl<'a> DynamicSystemData<'a> for BatchUncheckedWorld<'a> {
+    type Accessor = BatchAccessor;
+
+    fn setup(_accessor: &Self::Accessor, _world: &mut World) {}
+
+    fn fetch(_access: &Self::Accessor, world: &'a World) -> Self {
+        BatchUncheckedWorld(world)
+    }
+}
+
+/// The `BatchController` is the additional trait that a normal System must
+/// implement in order to be used as `BatchControllerSystem`.
+pub trait BatchController<'a, 'b> {
+    /// Here you must set all the resource types that you want to use inside the
+    /// `BatchControllerSystem`.
+    /// All the system data needed for the controlled systems is fetched
+    /// automatically.
+    type BatchSystemData: SystemData<'a>;
```

Comment:
Is there a reason this trait isn't `BatchController<'a, 'b>: System<'a>` and just use `<Self as System<'a>>::SystemData`?

Reply:
The `BatchControllerSystem` may want to read / write directly some resources.

Since it always has the `BatchUncheckedWorld`, I need a way to specify which resource it will use directly.

So the `BatchSystemData` serve to specify the resource that the controller will use.

Reply:
Ah now I understand what you are saying... But I am still not sure why cannot you specify `BatchControllerSystem` resources in the `System` `impl`? There has to be a reason, but it's not apparent from the code or documentation.

Reply:
Consider this code:
```
impl<'a> System<'a> for CustomBatchControllerSystem<'_, '_> {
    type SystemData = BatchUncheckedWorld<'a>;

    fn run(&mut self, data: Self::SystemData) {
        {
            // Drop before dispatch the inner dispatcher
            let _ts = data.0.fetch::<TomatoStore>();
        }
        println!("Batch execution");
        for _i in 0..3 {
            self.dispatcher.dispatch(data.0);
        }
    }
}
```
The main reason is that before to execute the inner dispatcher you want to drop all the fetched resources otherwise if even one of that resources will be fetched by the inner dispatcher it will panics.

The example already show how to use it, and write this implementation detail doesn't seems important for the "user" of the library.
What do you think?

Reply:
Thank you for explaining this in detail! Now I think I understand the relation between these two. I think it needs more explanation: If I wanted to implement my own batch controller I think I would have confused with this which would have made it harder than necessary.

Maybe add a comment that says that if the child systems depend on the same things as the batch controller there would be panic without separating these resource requirements.

Reply:
I've added this to the `BatchSystemData` what do you think?

    /// Here you must set all the `Resource` types that you want to use directly
    /// inside the `BatchControllerSystem`.
    /// 
    /// These have to be specified here, instead of `SystemData` (as
    /// a normal `System` does) because there is a direct dependecy with the sub `System`s
    /// `Resource`.
    /// 
    /// Since the sub `System`s will use the same `Resource' of the `BatchControllerSystem`
    /// is necessary to drop the references to the fetched `Resource` before to
    /// dispatch the inner `Dispatcher`.
    /// 
    /// Now is easy to understand that specify the `BatchControllerSystem` `Resource`
    /// in the `SystemData` doesn't allow to drop the reference before the sub dispatching
    /// resulting in `Panic`.
    /// 
    /// So this mechanism allow you to specify the `Resource` that you will use in the `BatchControllerSystem`
    /// then is safe to fetch them directly from the world (As the example  "examples/batch_dispatching.rs" show).
    ///
    /// Note that it's not required to specify the sub systems resources here because they
    /// are handled automatically.
    type BatchSystemData: SystemData<'a>;